proto/api_v2/metricsquery.proto

```diff
@@ -0,0 +1,158 @@
+  // Optional. Default = 1h.
+  google.protobuf.Duration lookback = 2 [
+    (gogoproto.stdduration) = true,
+    (gogoproto.nullable) = true
+  ];
+
+  // step size is the duration between data points of the query results.
+  // For example, if set to 5s, the results would produce a data point every 5 seconds
+  // from the start_time to end_time.
+  // Optional. Default = 5s.
+  google.protobuf.Duration step = 3 [
+    (gogoproto.stdduration) = true,
+    (gogoproto.nullable) = true
+  ];
+
+  // rate is the duration in which the per-second rate of change is calculated for a cumulative counter metric.
+  // Optional. Default = 10m.
+  google.protobuf.Duration rate = 4 [
+    (gogoproto.stdduration) = true,
+    (gogoproto.nullable) = true
+  ];
+}
+
+// GetMetricsResponse is the MetricsQueryService response returned by RPC calls.
+message GetMetricsResponse {
+  // metrics is the collection of metrics returned in a response.
+  repeated opentelemetry.proto.metrics.v1.Metric metrics = 1 [
+    (gogoproto.nullable) = false
+  ];
+}
+
+// GetServiceLatenciesRequest contains parameters for the GetServiceLatencies RPC call.
+message GetServiceLatenciesRequest {
+  MetricsQueryBaseRequest baseRequest = 1;
+  // quantile is the quantile to compute from latency histogram metrics.
+  // Required.
+  double quantile = 2;
+}
+
+// GetServiceCallRatesRequest contains parameters for the GetServiceCallRates RPC call.
+message GetServiceCallRatesRequest {
+  MetricsQueryBaseRequest baseRequest = 1;
+}
+
+// GetServiceErrorRatesRequest contains parameters for the GetServiceErrorRates RPC call.
+message GetServiceErrorRatesRequest {
+  MetricsQueryBaseRequest baseRequest = 1;
+}
+
+// GetOperationLatenciesRequest contains parameters for the GetOperationLatencies RPC call.
+message GetOperationLatenciesRequest {
+  MetricsQueryBaseRequest baseRequest = 1;
+  // service_name is the service name to fetch operation metrics from.
+  // Required.
+  string service_name = 2;
+  // quantile is the quantile to compute from latency histogram metrics.
+  // Required.
+  double quantile = 3;
+}
+
+// GetOperationCallRatesRequest contains parameters for the GetOperationCallRates RPC call.
+message GetOperationCallRatesRequest {
+  MetricsQueryBaseRequest baseRequest = 1;
+  // service_name is the service name to fetch operation metrics from.
+  // Required.
+  string service_name = 2;
+}
+
+// GetOperationErrorRatesRequest contains parameters for the GetOperationErrorRates RPC call.
+message GetOperationErrorRatesRequest {
+  MetricsQueryBaseRequest baseRequest = 1;
+  // service_name is the service name to fetch operation metrics from.
+  // Required.
+  string service_name = 2;
+}
+
+service MetricsQueryService {
+  // GetServiceLatencies gets the per-service latencies.
+  rpc GetServiceLatencies(GetServiceLatenciesRequest) returns (GetMetricsResponse) {
+    option (google.api.http) = {
+      get: "/metrics/latencies"
```

Comment:
I wouldn't bother with http, we're not using it anywhere.

Reply:
Well, you may need to try this out then. I tried that using grpc-gateway lib, I don't know if it's still the recommended way, and it was certainly weird back then, requiring you to run http server that would proxy the request to grpc server, even when they were in the same process.

Reply:
My mistake, I was confusing our native HTTP endpoints (port 16686) with the gRPC (port 16685) HTTP endpoints. Makes sense to not bother with HTTP on the gRPC listener; I'll remove it.